1033 Scopes On Mandatory Params

JChen2005 opened the pull request:
now allows the codgen to check for required scope on mandatory params

JChen2005 committed:
now allows the codgen to check for required scope on mandatory params

JChen2005 committed:
Added Scopes to all Params

abrown-RV reviewed (changes requested):
Allow for either `product` having the right scope or the individual params.

JChen2005 committed:
Added the option to also check the product struct's scope instead of just individual parameters.

JChen2005 committed:
Removed all scopes in the individual params and set it in the product struct.

JChen2005 committed:
Now allows the validator to see anything outside the param in the device model.

JChen2005 committed:
Removed extra lines at the bottom of each example files.

JChen2005 committed:
Added back newlines

abrown-RV commented on tools/codegen/index.js:
Should change the name and description of the param now that it is different.

JChen2005 replied:
I have fixed comments to include both param and scope.

JChen2005 replied:
Also changed param description to also reference the scopes.

abrown-RV replied:
Change the name of the param, its not the device params anymore, its the whole model.

abrown-RV commented on tools/codegen/index.js:
```diff
@@ -117,6 +120,17 @@ function areAllRequiredParamsPresent(deviceParams, disableMandatoryEnforcement =
 
+            // Validate the scope for mandatory parameters
+            const productScope = deviceParams.params.product.access_scope;
+            const globalScope = deviceParams.access_scopes || [];
```
Check against `default_scope`

abrown-RV commented on tools/codegen/index.js:
```diff
@@ -89,22 +89,25 @@ function areAllRequiredParamsPresent(deviceParams, disableMandatoryEnforcement =
     };
 
-    if (!deviceParams || !deviceParams.product) {
+    const REQUIRED_SCOPES = "st2138:mon";
```
not a plural

JChen2005 committed:
Added newline

abrown-RV commented on tools/codegen/index.js:
```diff
@@ -117,6 +120,17 @@ function areAllRequiredParamsPresent(deviceParams, disableMandatoryEnforcement =
+            const globalScope = deviceParams.access_scopes || [];
+            const hasGlobalRequiredScope = globalScope.includes(REQUIRED_SCOPES);
+            
+            if (param.access_scope && param.access_scope !== REQUIRED_SCOPES) {
```
I think it would be simpler to derive the scope for each param by combining the levels and then just a single if checking if its correct.

abrown-RV reviewed (changes requested):
still got a few more newlines to restore.

JChen2005 committed:
Added access_scopes to required structs in the devices, updated codegen to check for scopes

abrown-RV commented on tools/codegen/index.js:
```diff
@@ -136,7 +171,7 @@ function areAllRequiredParamsPresent(deviceParams, disableMandatoryEnforcement =
     });
 
-    const allIssues = [...missing.map(p => `${p} (missing field)`), ...emptyValues];
+        const allIssues = [...missing.map(p => `${p} (missing field)`), ...emptyValues, ...invalidScopes];
```
```suggestion
    const allIssues = [...missing.map(p => `${p} (missing field)`), ...emptyValues, ...invalidScopes];
```

JChen2005 replied:
I fixed the indenting.

abrown-RV reviewed (changes requested):
just about there, just a minor indentation fix. There's still an unresolved comment from last round as well.

JChen2005 committed:
Added Scope validator

JChen2005 committed:
Updated function comments

JChen2005 committed:
Added back newline

JChen2005 committed:
Adde back newline

NR-RV commented on tools/codegen/index.js:
```diff
@@ -117,6 +144,13 @@ function areAllRequiredParamsPresent(deviceParams, disableMandatoryEnforcement =
                 return;
             }
 
+            // Derive the effective scope and check if it's correct
```
check if this dose what you want it to do:

<img width="900" height="172" alt="image" src="https://github.com/user-attachments/assets/bbe7c554-f90b-4a0c-b664-02ef1e6f3661" />

here is a nice truth table you can fill out the outcome line with what your code is doing and see if that is what you want it to be doing

NR-RV replied:
here is an example of filling out the outcome (may not flow what your code dose)*:
<img width="938" height="432" alt="image" src="https://github.com/user-attachments/assets/6e12325f-7331-421d-8930-63d6c3c8475e" />

abrown-RV commented on tools/codegen/index.js:
```diff
@@ -89,22 +90,48 @@ function areAllRequiredParamsPresent(deviceParams, disableMandatoryEnforcement =
+            return defaultScope;
+        }
+
+        else {
```
Take this else out, its not needed

JChen2005 replied:
Redid the entire if statement, gave the else statment a use as it returns "Invalid" when other possible senarios are not possible.

JChen2005 committed:
Updated getDerivedScope function

JChen2005 committed:
Updated getDerivedScope

JChen2005 committed:
Edited function comment.

JChen2005 committed:
Updated checkParamAreValid function to
function validateRequiredParamsAndScopes.

JChen2005 committed:
Replaced arg deviceParams with deviceDesc and replaced each reference in function.

abrown-RV reviewed (approved):
Well done!

JChen2005 commented:
Closes issue #1033 

Updated function areAllRequiredParamsPresent by renaming it to validateRequiredParamsAndScopes.

Updated function comments.

Created a nested fucntion getDerivedScope to check and validate scopes in devices.
